lib/validators/spec/openapi.js

```diff
@@ -0,0 +1,244 @@
+ * @param {object}        path          - A Path object, from the Swagger API
+ * @param {string}        pathId        - A value that uniquely identifies the path
+ * @param {object}        operation     - An Operation object, from the Swagger API
+ * @param {string}        operationId   - A value that uniquely identifies the operation
+ */
+function validateParameters(api, path, pathId, operation, operationId) {
+  const pathParams = path.parameters || [];
+  const operationParams = operation.parameters || [];
+
+  // Check for duplicate path parameters.
+  try {
+    checkForDuplicates(pathParams);
+  } catch (e) {
+    throw ono.syntax(e, `Validation failed. ${pathId} has duplicate parameters`);
+  }
+
+  // Check for duplicate operation parameters.
+  try {
+    checkForDuplicates(operationParams);
+  } catch (e) {
+    throw ono.syntax(e, `Validation failed. ${operationId} has duplicate parameters`);
+  }
+
+  // Combine the path and operation parameters, with the operation params taking precedence over the path params.
+  const params = pathParams.reduce((combinedParams, value) => {
+    const duplicate = combinedParams.some(param => {
+      return param.in === value.in && param.name === value.name;
+    });
+
+    if (!duplicate) {
+      combinedParams.push(value);
+    }
+
+    return combinedParams;
+  }, operationParams.slice());
+
+  validatePathParameters(params, pathId, operationId);
+  validateParameterTypes(params, api, operation, operationId);
+}
+
+/**
+ * Validates path parameters for the given path.
+ *
+ * @param   {object[]}  params        - An array of Parameter objects
+ * @param   {string}    pathId        - A value that uniquely identifies the path
+ * @param   {string}    operationId   - A value that uniquely identifies the operation
+ */
+function validatePathParameters(params, pathId, operationId) {
+  // Find all `{placeholders}` in the path string. And because paths can have path parameters duped we need to convert
+  // this to a unique array so we can eliminate false positives of placeholders that might be duplicated.
+  const placeholders = [...new Set(pathId.match(util.swaggerParamRegExp) || [])];
+
+  params
+    .filter(param => param.in === 'path')
+    .forEach(param => {
+      if (param.required !== true) {
+        throw ono.syntax(
+          'Validation failed. Path parameters cannot be optional. ' +
+            `Set required=true for the "${param.name}" parameter at ${operationId}`
+        );
+      }
+
+      const match = placeholders.indexOf(`{${param.name}}`);
+      if (match === -1) {
+        throw ono.syntax(
+          `Validation failed. ${operationId} has a path parameter named "${param.name}", ` +
+            `but there is no corresponding {${param.name}} in the path string`
+        );
+      }
+
+      placeholders.splice(match, 1);
+    });
+
+  if (placeholders.length > 0) {
+    throw ono.syntax(`Validation failed. ${operationId} is missing path parameter(s) for ${placeholders}`);
```

Comment:
For all of these non-AJV validation cases we're showing error messages like this that I don't think are that great. I'd love to rework these into creating AJV-friendly errors that we can feed into our prettified code errors but I'm going to table that for now because it's a non-trivial amount of work.